// @vue/runtime-core/dist/runtime-core.esm.js
// packages/runtime-core/src/h.ts
import { isObject as isObject2 } from "@vue/reactivity";

// packages/runtime-core/src/createVnode.ts
import { ShapeFlags, normalizeClass } from "@vue/shared";
import { isString, isArray, isFunction, isObject } from "@vue/shared";
var Fragment = /* @__PURE__ */ Symbol("Fragment");
var Text = /* @__PURE__ */ Symbol("Text");
var Comment = /* @__PURE__ */ Symbol("Comment");
function isVnode(value) {
  return value.__v_isVnode === true;
}
function isVNode(value) {
  return value ? value.__v_isVNode === true : false;
}
function createVNode(type, props, children) {
  if (props) {
    let { class: klass, style } = props;
    if (klass && !isString(klass)) {
      props.class = normalizeClass(klass);
    }
  }
  const shapeFlag = isString(type) ? ShapeFlags.ELEMENT : isObject(type) ? ShapeFlags.STATEFUL_COMPONENT : 0;
  return createBaseVNode(type, props, children, shapeFlag);
}
function createBaseVNode(type, props, children, shapeFlag) {
  const vnode = {
    __v_isVNode: true,
    //标识
    type,
    props,
    shapeFlag,
    key: props?.key || null
  };
  normalizeChildren(vnode, children);
  return vnode;
}
function normalizeChildren(vnode, children) {
  let type = 0;
  const { shapeFlag } = vnode;
  if (children == null) {
    children = null;
  } else if (isArray(children)) {
    type = ShapeFlags.ARRAY_CHILDREN;
  } else if (typeof children === "object") {
  } else if (isFunction(children)) {
  } else {
    children = String(children);
    type = ShapeFlags.TEXT_CHILDREN;
  }
  vnode.children = children;
  vnode.shapeFlag |= type;
}
function isSameVNodeType(n1, n2) {
  return n1.type === n2.type && n1.key === n2.key;
}

// packages/runtime-core/src/h.ts
function h(type, propsOrChildren, children) {
  const l = arguments.length;
  if (l === 2) {
    if (isObject2(propsOrChildren) && !Array.isArray(propsOrChildren)) {
      if (isVnode(propsOrChildren)) {
        return createVNode(type, null, [propsOrChildren]);
      } else {
        return createVNode(type, propsOrChildren);
      }
    } else {
      return createVNode(type, null, propsOrChildren);
    }
  } else {
    if (l > 3) {
      children = Array.prototype.slice.call(arguments, 2);
    } else if (l === 3 && isVnode(children)) {
      children = [children];
    }
    return createVNode(type, propsOrChildren, children);
  }
}

// packages/runtime-core/src/render.ts
import { ShapeFlags as ShapeFlags3 } from "@vue/shared";
import { EMPTY_OBJ, isString as isString3 } from "@vue/shared";

// packages/runtime-core/src/componentRenderUtils.ts
function renderComponentRoot(instance) {
  const { vnode, render, data } = instance;
  let result;
  try {
    if (vnode.shapeFlag & 4 /* STATEFUL_COMPONENT */) {
      result = normalizeVNode(render.call(data));
    }
  } catch (err) {
    console.error(err);
  }
  return result;
}
function normalizeVNode(child) {
  if (typeof child === "object") {
    return cloneIfMounted(child);
  }
}
function cloneIfMounted(child) {
  return child;
}

// packages/runtime-core/src/component.ts
import { isFunction as isFunction2, isObject as isObject3 } from "@vue/shared";
import { reactive } from "@vue/reactivity";

// packages/runtime-core/src/apiLifecycle.ts
function injectHook(type, hook, target) {
  if (target) {
    target[type] = hook;
    return hook;
  }
}
var createHook = (lifecycle) => {
  return (hook, target) => injectHook(lifecycle, hook, target);
};
var onBeforeMount = createHook("bm" /* BEFORE_MOUNT */);
var onMounted = createHook("m" /* MOUNTED */);

// packages/runtime-core/src/component.ts
var uid = 0;
function createComponentInstance(vnode) {
  const type = vnode.type;
  const instance = {
    uid: uid++,
    // 唯一标记
    vnode,
    // 虚拟节点
    type,
    // 组件类型
    subTree: null,
    // render 函数的返回值
    effect: null,
    // ReactiveEffect 实例
    update: null,
    // update 函数，触发 effect.run
    render: null,
    // 组件内的 render 函数
    isMounted: false,
    // 是否挂载
    bc: null,
    // beforeCreate
    c: null,
    // created
    bm: null,
    // beforeMount
    m: null
    // mounted
  };
  return instance;
}
function applyOptions(instance) {
  const {
    data: dataOptions,
    beforeCreate,
    created,
    beforeMount,
    mounted
  } = instance.type;
  if (beforeCreate) {
    callHook(beforeCreate, instance.data);
  }
  if (dataOptions) {
    const data = dataOptions();
    if (isObject3(data)) {
      instance.data = reactive(data);
    }
  }
  if (created) {
    callHook(created, instance.data);
  }
  function registerLifecycleHook(register, hook) {
    register(hook?.bind(instance.data), instance);
  }
  registerLifecycleHook(onBeforeMount, beforeMount);
  registerLifecycleHook(onMounted, mounted);
}
function callHook(hook, proxy) {
  hook.bind(proxy)();
}
function setupComponent(instance) {
  const setupResult = setupStatefulComponent(instance);
  return setupResult;
}
function setupStatefulComponent(instance) {
  const Component = instance.type;
  const { setup } = Component;
  if (setup) {
    const setupResult = setup();
    handleSetupResult(instance, setupResult);
  } else {
    finishComponentSetup(instance);
  }
}
function handleSetupResult(instance, setupResult) {
  if (isFunction2(setupResult)) {
    instance.render = setupResult;
  }
  finishComponentSetup(instance);
}
function finishComponentSetup(instance) {
  const Component = instance.type;
  if (!instance.render) {
    instance.render = Component.render;
  }
  applyOptions(instance);
}

// packages/runtime-core/src/render.ts
import { ReactiveEffect } from "@vue/reactivity";

// packages/runtime-core/src/scheduler.ts
var isFlushPending = false;
var resolvedPromise = Promise.resolve();
var currentFlushPromise = null;
var pendingPreFlushCbs = [];
function queuePreFlushCb(cb) {
  queueCb(cb, pendingPreFlushCbs);
}
function queueCb(cb, pendingQueue) {
  pendingQueue.push(cb);
  queueFlush();
}
function queueFlush() {
  if (!isFlushPending) {
    isFlushPending = true;
    currentFlushPromise = resolvedPromise.then(flushJobs);
  }
}
function flushJobs() {
  isFlushPending = false;
  flushPreFlushCbs();
}
function flushPreFlushCbs() {
  if (pendingPreFlushCbs.length) {
    let activePreFlushCbs = [...new Set(pendingPreFlushCbs)];
    pendingPreFlushCbs.length = 0;
    for (let i = 0; i < activePreFlushCbs.length; i++) {
      activePreFlushCbs[i]();
    }
  }
}

// packages/runtime-core/src/getSequence.ts
function getSequence(arr) {
  const p = arr.slice();
  const result = [0];
  let i, j, u, v, c;
  const len = arr.length;
  for (i = 0; i < len; i++) {
    const arrI = arr[i];
    if (arrI !== 0) {
      j = result[result.length - 1];
      if (arr[j] < arrI) {
        p[i] = j;
        result.push(i);
        continue;
      }
      u = 0;
      v = result.length - 1;
      while (u < v) {
        c = u + v >> 1;
        if (arr[result[c]] < arrI) {
          u = c + 1;
        } else {
          v = c;
        }
      }
      if (arrI < arr[result[u]]) {
        if (u > 0) {
          p[i] = result[u - 1];
        }
        result[u] = i;
      }
    }
  }
  u = result.length;
  v = result[u - 1];
  while (u-- > 0) {
    result[u] = v;
    v = p[v];
  }
  return result;
}

// packages/runtime-core/src/render.ts
function createRenderer(rendererOptions) {
  const {
    insert: hostInsert,
    patchProp: hostPatchProp,
    createElement: hostCreateElement,
    setElementText: hostSetElementText,
    remove: hostRemove,
    createText: hostCreateText,
    setText: hostSetText,
    parentNode: hostParentNode,
    nextSibling: hostNextSibling,
    createComment: hostCreateComment
  } = rendererOptions;
  const patchProps = (el, vnode, oldProps, newProps) => {
    if (oldProps !== newProps) {
      for (const key in newProps) {
        const next = newProps[key];
        const prev = oldProps[key];
        if (next !== prev) {
          hostPatchProp(el, key, prev, next);
        }
      }
      if (oldProps !== EMPTY_OBJ) {
        for (const key in oldProps) {
          if (!(key in newProps)) {
            hostPatchProp(el, key, oldProps[key], null);
          }
        }
      }
    }
  };
  const processElement = (oldVNode, newVNode, container, anchor) => {
    if (oldVNode == null) {
      mountElement(newVNode, container, anchor);
    } else {
      patchElement(oldVNode, newVNode);
    }
  };
  const patchElement = (oldVNode, newVNode) => {
    const el = newVNode.el = oldVNode.el;
    const oldProps = oldVNode.props || EMPTY_OBJ;
    const newProps = newVNode.props || EMPTY_OBJ;
    patchChildren(oldVNode, newVNode, el, null);
    patchProps(el, newVNode, oldProps, newProps);
  };
  const patchChildren = (oldVNode, newVNode, container, anchor) => {
    const c1 = oldVNode && oldVNode.children;
    const prevShapeFlag = oldVNode ? oldVNode.shapeFlag : 0;
    const c2 = newVNode.children;
    const { shapeFlag } = newVNode;
    if (shapeFlag & ShapeFlags3.TEXT_CHILDREN) {
      if (prevShapeFlag & ShapeFlags3.ARRAY_CHILDREN) {
      }
      if (c2 !== c1) {
        hostSetElementText(container, c2);
      }
    } else {
      if (prevShapeFlag & ShapeFlags3.ARRAY_CHILDREN) {
        if (shapeFlag & ShapeFlags3.ARRAY_CHILDREN) {
          patchKeyedChildren(c1, c2, container, anchor);
        } else {
        }
      } else {
        if (prevShapeFlag & ShapeFlags3.TEXT_CHILDREN) {
          hostSetElementText(container, "");
        }
        if (shapeFlag & ShapeFlags3.ARRAY_CHILDREN) {
        }
      }
    }
  };
  const processText = (oldVNode, newVNode, container, anchor) => {
    if (oldVNode == null) {
      newVNode.el = hostCreateText(newVNode.children);
      hostInsert(newVNode.el, container, anchor);
    } else {
      const el = newVNode.el = oldVNode.el;
      if (newVNode.children !== oldVNode.children) {
        hostSetText(el, newVNode.children);
      }
    }
  };
  const processCommentNode = (oldVNode, newVNode, container, anchor) => {
    if (oldVNode == null) {
      newVNode.el = hostCreateComment(newVNode.children || "");
      hostInsert(newVNode.el, container, anchor);
    } else {
      newVNode.el = oldVNode.el;
    }
  };
  const processFragment = (oldVNode, newVNode, container, anchor) => {
    if (oldVNode == null) {
      mountChildren(newVNode.children, container);
    } else {
      patchChildren(oldVNode, newVNode, container, anchor);
    }
  };
  const processComponent = (oldVNode, newVNode, container, anchor) => {
    if (oldVNode == null) {
      mountComponent(newVNode, container, anchor);
      console.log("\u6302\u8F7D\u7EC4\u4EF6", newVNode);
    }
  };
  const patchKeyedChildren = (oldChildren, newChildren, container, parentAnchor) => {
    let i = 0;
    const newChildrenLength = newChildren.length;
    let oldChildrenEnd = oldChildren.length - 1;
    let newChildrenEnd = newChildrenLength - 1;
    while (i <= oldChildrenEnd && i <= newChildrenEnd) {
      const oldVNode = oldChildren[i];
      const newVNode = normalizeVNode(newChildren[i]);
      if (isSameVNodeType(oldVNode, newVNode)) {
        patch(oldVNode, newVNode, container, null);
      } else {
        break;
      }
      i++;
    }
    while (i <= oldChildrenEnd && i <= newChildrenEnd) {
      const oldVNode = oldChildren[oldChildrenEnd];
      const newVNode = normalizeVNode(newChildren[newChildrenEnd]);
      if (isSameVNodeType(oldVNode, newVNode)) {
        patch(oldVNode, newVNode, container, null);
      } else {
        break;
      }
      oldChildrenEnd--;
      newChildrenEnd--;
    }
    if (i > oldChildrenEnd) {
      if (i <= newChildrenEnd) {
        const nextPos = newChildrenEnd + 1;
        const anchor = nextPos < newChildrenLength ? newChildren[nextPos].el : parentAnchor;
        while (i <= newChildrenEnd) {
          patch(null, normalizeVNode(newChildren[i]), container, anchor);
          i++;
        }
      }
    } else if (i > newChildrenEnd) {
      while (i <= oldChildrenEnd) {
        unmount(oldChildren[i]);
        i++;
      }
    } else {
      const oldStartIndex = i;
      const newStartIndex = i;
      const keyToNewIndexMap = /* @__PURE__ */ new Map();
      for (i = newStartIndex; i <= newChildrenEnd; i++) {
        const nextChild = normalizeVNode(newChildren[i]);
        if (nextChild.key != null) {
          keyToNewIndexMap.set(nextChild.key, i);
        }
      }
      let j;
      let patched = 0;
      const toBePatched = newChildrenEnd - newStartIndex + 1;
      let moved = false;
      let maxNewIndexSoFar = 0;
      const newIndexToOldIndexMap = new Array(toBePatched);
      for (i = 0; i < toBePatched; i++) newIndexToOldIndexMap[i] = 0;
      for (i = oldStartIndex; i <= oldChildrenEnd; i++) {
        const prevChild = oldChildren[i];
        if (patched >= toBePatched) {
          unmount(prevChild);
          continue;
        }
        let newIndex;
        if (prevChild.key != null) {
          newIndex = keyToNewIndexMap.get(prevChild.key);
        }
        if (newIndex === void 0) {
          unmount(prevChild);
        } else {
          newIndexToOldIndexMap[newIndex - newStartIndex] = i + 1;
          if (newIndex >= maxNewIndexSoFar) {
            maxNewIndexSoFar = newIndex;
          } else {
            moved = true;
          }
          patch(prevChild, newChildren[newIndex], container, null);
          patched++;
        }
      }
      const increasingNewIndexSequence = moved ? getSequence(newIndexToOldIndexMap) : [];
      j = increasingNewIndexSequence.length - 1;
      for (i = toBePatched - 1; i >= 0; i--) {
        const nextIndex = newStartIndex + i;
        const nextChild = newChildren[nextIndex];
        const anchor = nextIndex + 1 < newChildrenLength ? newChildren[nextIndex + 1].el : parentAnchor;
        if (newIndexToOldIndexMap[i] === 0) {
          patch(null, nextChild, container, anchor);
        } else if (moved) {
          if (j < 0 || i !== increasingNewIndexSequence[j]) {
            move(nextChild, container, anchor);
          } else {
            j--;
          }
        }
      }
    }
  };
  const move = (vnode, container, anchor) => {
    const { el } = vnode;
    hostInsert(el, container, anchor);
  };
  const mountComponent = (initialVNode, container, anchor) => {
    initialVNode.component = createComponentInstance(initialVNode);
    const instance = initialVNode.component;
    setupComponent(instance);
    setupRenderEffect(instance, initialVNode, container, anchor);
  };
  const setupRenderEffect = (instance, initialVNode, container, anchor) => {
    const componentUpdateFn = () => {
      if (!instance.isMounted) {
        const { bm, m } = instance;
        if (bm) {
          bm();
        }
        const subTree = instance.subTree = renderComponentRoot(instance);
        patch(null, subTree, container, anchor);
        if (m) {
          m();
        }
        initialVNode.el = subTree.el;
        instance.isMounted = true;
      } else {
        let { next, vnode } = instance;
        if (!next) {
          next = vnode;
        }
        const nextTree = renderComponentRoot(instance);
        const prevTree = instance.subTree;
        instance.subTree = nextTree;
        patch(prevTree, nextTree, container, anchor);
        next.el = nextTree.el;
      }
    };
    const effect = instance.effect = new ReactiveEffect(
      componentUpdateFn,
      () => queuePreFlushCb(update)
    );
    const update = instance.update = () => effect.run();
    update();
  };
  const unmount = (vnode) => {
    hostRemove(vnode.el);
  };
  const mountChildren = (children, container, anchor) => {
    if (isString3(children)) {
      children = children.split("");
    }
    for (let i = 0; i < children.length; i++) {
      const child = children[i] = normalizeVNode(children[i]);
      patch(null, child, container, anchor);
    }
  };
  const mountElement = (vnode, container, anchor = null) => {
    const { type, children, props, shapeFlag } = vnode;
    const el = hostCreateElement(type);
    vnode.el = el;
    if (props) {
      for (const key in props) {
        hostPatchProp(el, key, null, props[key]);
      }
    }
    if (shapeFlag & ShapeFlags3.TEXT_CHILDREN) {
      hostSetElementText(el, children);
    } else if (shapeFlag & ShapeFlags3.ARRAY_CHILDREN) {
      mountChildren(children, el, anchor);
    }
    hostInsert(el, container);
  };
  const patch = (n1, n2, container, anchor = null) => {
    if (n1 && !isSameVNodeType(n1, n2)) {
      unmount(n1);
      n1 = null;
    }
    if (n1 == n2) {
      return;
    }
    const { type, shapeFlag } = n2;
    switch (type) {
      case Text:
        processText(n1, n2, container, anchor);
        break;
      case Comment:
        processCommentNode(n1, n2, container, anchor);
        break;
      case Fragment:
        processFragment(n1, n2, container, anchor);
        break;
      default:
        if (shapeFlag & ShapeFlags3.ELEMENT) {
          processElement(n1, n2, container, anchor);
        } else if (shapeFlag & ShapeFlags3.COMPONENT) {
          processComponent(n1, n2, container, anchor);
        }
    }
  };
  const render = (vnode, container) => {
    if (vnode == null) {
      if (container._vnode) {
        unmount(container._vnode);
      }
    } else {
      patch(container._vnode || null, vnode, container);
    }
    container._vnode = vnode;
  };
  return {
    render
  };
}
export {
  Comment,
  Fragment,
  Text,
  createRenderer,
  createVNode,
  h,
  isSameVNodeType,
  isVNode,
  isVnode,
  normalizeChildren
};
